test/gtest/ucp/test_ucp_tag_cancel.cc

```diff
@@ -32,4 +44,42 @@ UCS_TEST_P(test_ucp_tag_cancel, cancel_exp) {
     request_release(req);
 }
 
+// Test that cancelling already matched (but not yet completed) request does
+// not produce any error. GH bug #4490.
+UCS_TEST_P(test_ucp_tag_cancel, cancel_matched, "RNDV_THRESH=32K") {
+    uint64_t small_data = 0;
+    ucp_tag_t tag       = 0xfafa;
+    size_t size         = 50000;
+
+    std::vector<char> sbuf(size, 0);
+    std::vector<char> rbuf(size, 0);
+
+    request *rreq1 = recv_nb(&rbuf[0], rbuf.size(), DATATYPE, tag,
+                             UCP_TAG_MASK_FULL);
+    request *rreq2 = recv_nb(&small_data, sizeof(small_data), DATATYPE, tag,
+                             UCP_TAG_MASK_FULL);
+
+    request *sreq1 = send_nb(&sbuf[0], sbuf.size(), DATATYPE, tag);
+    request *sreq2 = send_nb(&small_data, sizeof(small_data), DATATYPE, tag);
+
+    wait(rreq2);
+    EXPECT_EQ(UCS_OK, rreq2->status);
+
+    if (!rreq1->completed) {
+        ucp_request_cancel(receiver().worker(), rreq1);
+    } else {
+        UCS_TEST_MESSAGE << "nothing to cancel";
+    }
+
+    wait(rreq1);
```

Comment:
you can rename `release_send_req()` -> `release_req()` (or smth like that if it's already used by another function) and use it for `rreq1`. it should decrease code duplication, since you did the same for send requests

BTW, what if you call new added function `release_req()` for `rreq2`? Is it not so bad to do `wait(rreq2)` twice?